slides/02_traits.md

```diff
@@ -0,0 +1,376 @@
+- Notification: events
+- Visualization: MVC, automatic GUI!
+
+<!-- #endregion -->
+
+<!-- #region slideshow={"slide_type": "slide"} -->
+## An example
+
+<!-- #endregion -->
+
+```python
+from traits.api import Delegate, HasStrictTraits, Instance, Int, Str, observe
+
+class Parent(HasStrictTraits):
+    # INITIALIZATION: 'last_name' initialized to ''
+    last_name = Str('')
+
+```
+
+```python
+class Child(HasStrictTraits):
+    age = Int
+    # VALIDATION: 'father' must be Parent instance
+    father = Instance(Parent)
+    # DELEGATION: 'last_name' delegated to father's
+    last_name = Delegate('father')
+    # NOTIFICATION: Method called when 'age' changes
+    def _age_changed(self, old, new):
+        print('Age changed from %s to %s ' % (old, new))
+
+```
+
+<!-- #region slideshow={"slide_type": "slide"} -->
+## Using this
+
+<!-- #endregion -->
+
+```python
+joe = Parent()
+joe.last_name = 'Johnson'
+moe = Child()
+moe.father = joe
+```
+
+```python
+# Delegation
+moe.last_name
+```
+
+```python
+# Notification
+moe.age = 10
+```
+
+```python
+# Validation
+moe.age = '1'
+```
+
+```python
+# Visualization
+moe.configure_traits()
+```
+
+- Live editing!
+
+```python
+%gui qt
+```
+
+```python
+moe.edit_traits()
+```
+
+```python
+moe.age = 21
+```
+
+<!-- #region slideshow={"slide_type": "slide"} -->
+## Predefined trait types
+
+- Standard: `Bool, Complex, Int, Float, Str, Tuple, List, Dict`
+- Constrained: `Range, Regex, Expression, ReadOnly`
+- Special: `Date, Either/Union, Enum, Array, File, Color, Font, Button`
+- Generic: `Instance, Any, Callable`
+- Custom traits: 2D/3D plots etc.
+
+<!-- #endregion -->
+
+<!-- #region slideshow={"slide_type": "slide"} -->
+## `HasStrictTraits` vs. `HasStrictTraits`
+
+- Better to use `HasStrictTraits`
+- Will catch errors when you mistype or misspell an attribute
+- Will not allow setting any attribute not already declared
+
+
+<!-- #endregion -->
+
+
+<!-- #region slideshow={"slide_type": "slide"} -->
+## Notification example
+
+<!-- #endregion -->
+
+```python
+class Parent(HasStrictTraits):
+    last_name = Str('')
+
+
+class Child(HasStrictTraits):
+    age = Int
+    father = Instance(Parent)
+
+    def _age_changed(self, old, new):
+        print('Age changed from %s to %s ' % (old, new))
+
+    @observe('father.last_name')
+    def _dad_name_updated(self, event):
+        print('DAD name', self.father.last_name)
+
+```
+
+```python
+dad = Parent(last_name='Zubizaretta')
+c = Child(father=Parent)
+```
+
+```python
+dad.last_name = 'Valderrama'
+```
+
+```python
+def handler(event):
+    print("handler", event.object, event.name, event.old, event.new)
+```
+
+```python
+c = Child(father=Parent(last_name='Ram'))
+c.observe(handler, 'father, age')
+```
+
+<!-- #region slideshow={"slide_type": "slide"} -->
+## Trait change notification
+
+- Static: `def _<trait_name>_changed()`
+- Decorator: `@observe('extended.trait.name')`
+- Dynamic:
+
+```obj.observe(handler, 'extended.trait.name')
+```
```

Comment:
This is a topic we could shelve if we think we're running short.

Reply:
I would just introduce the decorator so I agree if that's what you mean.

Reply:
Yup. Again, dynamically adding and removing listeners isn't something a beginner or even intermediate developer would need.

Reply:
I can dump this part for sure.